File: src/main/python/AoC2015_08.py
#! /usr/bin/env python3
#
# Advent of Code 2015 Day 8
#

import re
import sys
from enum import Enum
from enum import auto
from enum import unique

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

TEST = r"""""
"abc"
"aaa\"aaa"
"\x27"
"""
RE = re.compile(r"\\x[0-9a-f]{2}")

Input = InputData
Output1 = int
Output2 = int


@unique
class Mode(Enum):
    DECODE = auto()
    ENCODE = auto()

    def overhead(self, string: str) -> int:
        match self:
            case Mode.DECODE:
                assert string[0] == '"' and string[-1] == '"'
                cnt = 2
                while string.find(r"\\") != -1:
                    string = string.replace(r"\\", "", 1)
                    cnt += 1
                return cnt + string.count(r"\"") + 3 * len(RE.findall(string))
            case Mode.ENCODE:
                return 2 + string.count("\\") + string.count('"')


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return input_data

    def solve(self, inputs: Input, mode: Mode) -> int:
        return sum(mode.overhead(s) for s in inputs)

    def part_1(self, inputs: Input) -> Output1:
        return self.solve(inputs, Mode.DECODE)

    def part_2(self, inputs: Input) -> Output2:
        return self.solve(inputs, Mode.ENCODE)

    @aoc_samples(
        (
            ("part_1", TEST, 12),
            ("part_2", TEST, 19),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2015, 8)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2015_08.java
import static com.github.pareronia.aoc.StringUtils.countMatches;

import java.util.List;
import java.util.regex.Pattern;

import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2015_08 extends SolutionBase<List<String>, Integer, Integer> {
    
    private static final Pattern HEX = Pattern.compile("\\\\x[0-9a-f]{2}");

    private AoC2015_08(final boolean debug) {
        super(debug);
    }

    public static final AoC2015_08 create() {
        return new AoC2015_08(false);
    }

    public static final AoC2015_08 createDebug() {
        return new AoC2015_08(true);
    }

    @Override
    protected List<String> parseInput(final List<String> inputs) {
        return inputs;
    }

    private int solve(final List<String> inputs, final Mode mode) {
        return inputs.stream().mapToInt(mode::overhead).sum();
    }

    @Override
    public Integer solvePart1(final List<String> inputs) {
        return solve(inputs, Mode.DECODE);
    }

    @Override
    public Integer solvePart2(final List<String> inputs) {
        return solve(inputs, Mode.ENCODE);
    }

    private enum Mode {
        DECODE, ENCODE;

        public int overhead(String str) {
            return switch (this) {
                case DECODE -> {
                    assert str.charAt(0) == '"' && str.charAt(str.length() - 1) == '"';
                    int cnt = 2;
                    while (str.contains("\\\\")) {
                        str = str.replaceFirst("[\\\\]{2}", "");
                        cnt++;
                    }
                    cnt += countMatches(str, "\\\"");
                    cnt += 3 * HEX.matcher(str).results().count();
                    yield cnt;
                }
                case ENCODE ->
                        2 + countMatches(str, "\\") + countMatches(str, "\"");
            };
        }
    }

    @Override
    @Samples({
        @Sample(method = "part1", input = TEST, expected = "12"),
        @Sample(method = "part2", input = TEST, expected = "19"),
    })
    public void samples() {
    }

    public static void main(final String[] args) throws Exception {
        AoC2015_08.create().run();
    }

    private static final String TEST =
            """
            ""
            "abc"
            "aaa\\"aaa"
            "\\x27\"\
            """;
}
